src/viam/sdk/common/proto_value.hpp

```diff
@@ -0,0 +1,324 @@
+
+    ProtoValue(const ProtoValue& other);
+
+    /// @brief Move assignment from other, leaving other in its unspecified-but-valid moved from
+    /// state.
+    ProtoValue& operator=(ProtoValue&& other) noexcept(impl::all_proto_moves_noexcept{});
+
+    ProtoValue& operator=(const ProtoValue& other);
+
+    ~ProtoValue();
+
+    /// @brief Test equality of two types.
+    /// @note "Intuitive" arithmetic equality is not supported, but could be.
+    /// Thus, bool{false}, int{0}, and double{0.0} do not compare equal.
+    friend bool operator==(const ProtoValue& lhs, const ProtoValue& rhs);
+
+    void swap(ProtoValue& other) noexcept(impl::all_proto_moves_noexcept{});
+
+    /// @brief Construct from proto value
+    /// @note This method is trivially templated to insulate google::protobuf::Value from our
+    /// API/ABI. It is meant to be called with no template parameters in a translation unit which
+    /// includes <google/protobuf/struct.pb.h>
+    template <typename Value = google::protobuf::Value>
+    static ProtoValue from_proto(const Value& v);  // NOLINT(misc-no-recursion)
+
+    friend void to_proto(const ProtoValue& t, google::protobuf::Value* v);
+
+    /// @name Casting API
+    ///@{
+
+    /// @brief Obtain integer constant representing the stored data type.
+    int kind() const;
+
+    /// @brief Checks whether this ProtoT is an instance of type T.
+    template <typename T>
+    bool is_a() const;
+
+    /// @brief Checking cast to T, returns non-owning non-null pointer if argument is_a<T>()
+    template <typename T>
+    friend T* dyn_cast(ProtoValue&);
+
+    /// @brief Checking cast to T, returns non-owning non-null pointer if argument is_a<T>()
+    template <typename T>
+    friend T const* dyn_cast(const ProtoValue&);
+
+    ///@}
+
+   private:
+    // This struct is our implementation of a virtual table, similar to what is created by the
+    // compiler for polymorphic types. We can't use actual polymorphic types because the
+    // implementation uses aligned stack storage, so we DIY it insted.
+    // The vtable can be thought of as defining a concept or interface which our ProtoValue types
+    // must satisfy.
+    // The first void [const]* parameter in any of the pointers below is always the `this` or `self`
+    // pointer.
+    struct vtable {
+        void (*dtor)(void*);
+        void (*copy)(void const*, void*);
+        void (*move)(void*, void*);
+        void (*to_proto)(void const*, google::protobuf::Value*);
+        int (*kind)();
+        bool (*equal_to)(void const*, void const*, const vtable&);
+    };
+
+    // Class template for a concrete model of the concept described by vtable.
+    // All construction and equality operations are implemented using defaults of the stored T.
+    template <typename T>
+    struct model {
+        model(T t) noexcept(std::is_nothrow_move_constructible<T>{});
+
+        static void dtor(void* self) noexcept;
```

Comment:
Do you want a static assertion that T is nothrow destructible?

Reply:
Not sure if it's necessary? In our closed universe of types that's always true, and also throwing destructors are UB anyway right

Reply:
Well, the `static_assert` doesn't cost much. It is possible to have `noexcept(false)` dtors. We don't have any of those right now I believe, and I think it is unlikely that we would. But I think I'd prefer to know if somehow one materialized unexpectedly in the future.